File: Codes/1483-kth-ancestor-of-a-tree-node.py
'''
Author: LetMeFly
Date: 2024-04-06 12:40:18
LastEditors: LetMeFly
LastEditTime: 2024-04-06 12:45:59
'''
from typing import List


Log = 16

class TreeAncestor:
    def __init__(self, n: int, parent: List[int]):
        self.ancestors = [[parent[i]] + [-1] * (Log - 1) for i in range(n)]
        for j in range(1, Log):
            for i in range(n):
                if self.ancestors[i][j - 1] != -1:
                    self.ancestors[i][j] = self.ancestors[self.ancestors[i][j - 1]][j - 1]

    def getKthAncestor(self, node: int, k: int) -> int:
        for j in range(Log):
            if (k >> j) & 1:
                node = self.ancestors[node][j]
            if node == -1:
                break
        return node


# Your TreeAncestor object will be instantiated and called as such:
# obj = TreeAncestor(n, parent)
# param_1 = obj.getKthAncestor(node,k)

File: Codes/1483-kth-ancestor-of-a-tree-node.cpp
/*
 * @Author: LetMeFly
 * @Date: 2024-04-06 12:29:13
 * @LastEditors: LetMeFly
 * @LastEditTime: 2024-04-06 12:39:33
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class TreeAncestor {
private:
    const static int Log = 16;  // 2 ^ 16 = 65536
    vector<vector<int>> ancestors;
public:
    TreeAncestor(int n, vector<int>& parent) {
        ancestors = vector<vector<int>>(n, vector<int>(Log, -1));
        for (int i = 0; i < n; i++) {
            ancestors[i][0] = parent[i];
        }
        for (int j = 1; j < Log; j++) {
            for (int i = 0; i < n; i++) {
                if (ancestors[i][j - 1] != -1) {  // don't forget
                    ancestors[i][j] = ancestors[ancestors[i][j - 1]][j - 1];
                }
            }
        }
    }
    
    int getKthAncestor(int node, int k) {
        for (int j = 0; j < Log && node != -1; j++) {
            if ((k >> j) & 1) {
                node = ancestors[node][j];
            }
        }
        return node;
    }
};

/**
 * Your TreeAncestor object will be instantiated and called as such:
 * TreeAncestor* obj = new TreeAncestor(n, parent);
 * int param_1 = obj->getKthAncestor(node,k);
 */